### PR TITLE
chore(runner,runner-ct): reduce duplication with packages/runner-shared

### PR DESCRIPTION
**Pull request by lmiller1990**
Context and initial proposal (Cypress internal team only): https://www.notion.so/cypressdx/Sharing-Code-E2E-CT-cf639ecb1f5e4c4dbc301ee0490681a5

- created `@packages/runner-shared`
- shared many common UI components
- shared common logic (`event-manager.js`, etc)
- move all common code into `@packages/shared-runner`

Although there is around 100 files changed, the changes themselves are minimal, as you will see. No new features and minimal code changes were made. I tried to do this with a little destruction to the git history by using `git mv`.

There is some more sharing we need to do, between `server` and `server-ct`. This PR is already quite large, so I'll make those changes in a separate one.

### CODE REVIEW

File: packages/runner-ct/src/app/ReporterContainer.tsx [hunk 1]
```diff
@@ -15,7 +13,10 @@ import styles from './RunnerCt.module.scss'
 interface ReporterContainerProps {
   state: State
   eventManager: typeof EventManager
-  config: Cypress.RuntimeConfigOptions
```
Comment (lmiller1990) [inline packages/runner-ct/src/app/ReporterContainer.tsx:18]: I am really not sure what the correct type definition is for this. It's certainly not `Cypress.RuntimeConfigOptions`, though, there are some properties that definitely exist that are not included in that type. Or, the type is wrong.

File: packages/runner-shared/src/container/container.spec.jsx [hunk 2]
```diff
@@ -117,43 +136,19 @@ describe('<Container />', () => {
...
-      props.util.hasSpecFile.returns(true)
+      props.hasSpecFile.returns(true)
       component.find(NoSpec).prop('onHashChange')()
       component.update()
       expect(component.find(App)).to.exist
     })
   })
-
-  describe('when automation is connected and there is a spec file', () => {
-    let props
```
Comment (lmiller1990) [inline packages/runner-shared/src/container/container.spec.jsx:141]: This test was not doing anything very useful and was very brittle so I just nuked it 🤷

File: packages/runner-shared/src/container/index.jsx [hunk 3]
```diff
@@ -82,18 +79,23 @@ class Container extends Component {
...
-  util,
-}
+  // This two functions, _noSpec anad _checkSpecFile, are only used by the E2E runner.
+  // TODO: remove any runner specific code from this file.
+  _noSpec () {
+    const { NoSpec } = this.props
 
-Container.propTypes = {
-  config: PropTypes.object.isRequired,
-  state: PropTypes.instanceOf(State),
```
Comment (lmiller1990) [inline packages/runner-shared/src/container/index.jsx:94]: This was tricky, since the `Container.jsx` is now shared, it's not trivial to import `State` and use it in the propType validation. What we should really do is change this to TypeScript, then we could use a generic type to validate this.

I tried to avoid changing `jsx` to `tsx` for anything but the most simple files in this PR to keep it as small as possible, so I think this is the most practical option. In my experience `propTypes` is not super useful anyway, and our tests are much more thorough and give us enough coverage.

File: packages/runner-shared/src/event-manager.js [hunk 4]
```diff
@@ -289,10 +298,15 @@ const eventManager = {
...
     // expose Cypress globally
-    window.Cypress = Cypress
+    // since CT AUT shares the window with the spec, we don't want to overwrite
+    // our spec Cypress instance with the component's Cypress instance
+    if (window.top === window) {
+      window.Cypress = Cypress
+    }
 
-    this._addListeners()
+    this._addListeners(Cypress)
```
Comment (lmiller1990) [inline packages/runner-shared/src/event-manager.js:309]: Not entirely sure what this does. `event-manager.js` had some ever so slight differences in E2E and CT, I took a ton  of time and care to merge them. This doesn't seem to break anything - anyone have some more context? Happy to dive deeper if anyone senses some risk.

File: packages/runner-shared/src/header/index.tsx [hunk 5]
```diff
@@ -0,0 +1,242 @@
+import cs from 'classnames'
+
+import { action, computed, observable } from 'mobx'
+import { observer } from 'mobx-react'
+import React, { Component, createRef } from 'react'
+import Tooltip from '@cypress/react-tooltip'
+// @ts-ignore - yes it does!
+import { $ } from '@packages/driver'
```
Comment (lmiller1990) [inline packages/runner-shared/src/header/index.tsx:8]: TS thinks that `$` is not exported from `@packages/driver`, but actually it is.

File: packages/runner-ct/src/app/SpecContent.tsx [hunk 6]
```diff
@@ -1,27 +1,27 @@
...
-import { namedObserver } from '../lib/mobx'
 
 interface SpecContentProps {
   state: State
   eventManager: typeof EventManager
-  config: Cypress.RuntimeConfigOptions
+  config: {
+    configFile: string
+    [key: string]: unknown
+  }
```
Comment (flotwig) [inline packages/runner-ct/src/app/SpecContent.tsx:24]: This looks like a step backwards with types, is it possible to get Cypress. types here?

File: packages/runner-shared/src/header/index.tsx (hunk 5 above)
Comment (flotwig) [reply packages/runner-shared/src/header/index.tsx:8]: I think you could update `packages/driver/index.d.ts` to correctly reflect the export

File: packages/runner-shared/src/event-manager.js [hunk 7]
```diff
@@ -289,10 +298,15 @@ const eventManager = {
   setup (config) {
     Cypress = this.Cypress = $Cypress.create(config)
 
+    // TODO: this was CT only before. Does this make sense for e2e too?
```
Comment (flotwig) [inline packages/runner-shared/src/event-manager.js:301]: it doesn't seem to be logically different, based on the fact that the tests still pass, so imma say "yes"

```suggestion
```

File: packages/runner-shared/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+export * from './viewport-info'
+
+export * from './config-file-formatted'
+
+export * from './header'
+
+export * from './container'
+
+export * from './automation-element'
```
Comment (flotwig) [inline packages/runner-shared/src/index.ts:41]: I'm not sure I agree with how these separate namespaces are all jumbled together and exported, I don't understand if there is a good reason for this.

You can't do `export * as Foo from './foo'`, but you can do it in two steps, like in `packages/network/lib/index.ts`: https://github.com/cypress-io/cypress/blob/77880f1a0f0cecaee92e1f271cede452cc133287/packages/network/lib/index.ts#L1-L19

@lmiller1990 can this be updated so this so that it's namespaced by module instead of mixed up?

**Review by flotwig — CHANGES_REQUESTED** [review]
This is fantastic Lachlan, I think it's almost ready to merge.

File: packages/runner-ct/src/app/SpecContent.tsx (hunk 6 above)
Comment (flotwig) [reply packages/runner-ct/src/app/SpecContent.tsx:24]: oh, i assume this is related to your other comment re: runtimeconfigoptions

**Review by jennifer-shehane — DISMISSED** [review]
@lmiller1990 There seems to be some differences in the percy snapshots that indicate some of the css has unintentionally changed. Let's make sure there are no changes in there.

File: packages/runner-shared/src/index.ts (hunk 8 above)
Comment (lmiller1990) [reply packages/runner-shared/src/index.ts:41]: This is (mostly) set up in a similar manner. There is two main scenarios to consider: 

## 1. single files (inside of `packages/runner-shared/src`). Examples are `dom.js` and `logger.js`. 

Each root level file has a single named export - for example:
- 
- Each of those exports is either a file (eg `packages/runner-shared/dom.js`) with a named export:

```js
// packages/runner-shared/src/dom.js

/* implementation */

export const dom = {
  addBlackout,
  removeBlackouts,
  addElementBoxModelLayers,
  // many more functions
}
```

So when you do `export * from './dom'`, it's just the same as what you are suggesting - except instead of import and re-exporting, we just do it in one line.

Usage is there:

```js
import { dom } from '@packages/runner-shared'`

dom.addBlackout()
```

So it is still namespaced. This same pattern is applied to all root level files.

## 2. Directories with multiple files

The other scenario is a directory, like `snapshot-controls`. It has three files:

```jsx
- index.js
- snapshot-controls.jsx
- snapshot-controls.spec.jsx
```

`index.js` simply exposes the module interface, in this case it does `export * as from ./snapshot-controls'`. It only contains 1 named export, `<SnapshotControls />` so name-spacing it doesn't make much sense. We could do `<Snapshot.Controls />`, but it seems overkill (a `Snapshot` namespace for a single component?) 

Usage is:

```
import { SnapshotControls, dom, logger } from 'packages/runner-shared`

So it's pretty well namespaced.

## 3. Exception

The only exception is the `studio` directory, which two components (`<Studio>` and `<StudioModals>`) and one singleton (`studioModel`). We could namespace these, but it feels a bit odd to have React components and a global singleton in the same namespace:

```jsx
import { Studio } from '@packages/runner-shared'

export function SomeComponent () {
  studio.studioModel(/* stuff */)
  
  return (
    <studio.StudioModals />
  )
}
}
```

What do you think?

File: packages/runner-shared/src/header/index.tsx (hunk 5 above)
Comment (lmiller1990) [reply packages/runner-shared/src/header/index.tsx:7]: Done - did not realize this was typed with a `.d.ts` file.

**Comment by lmiller1990** [comment]
@flotwig thanks for the review. Responded to your comment, happy to make further changes if you think they make sense.

@jennifer-shehane I found and fixed the visual regression. There's now only 1 difference, all in RunnerCt (take a look). This change is desirable, RunnerCt and Runner had slight differences, now they are using the same component, the Percy snapshot needs updating. I'll approve the Percy build as the last thing before merging this up.

File: packages/driver/index.d.ts [hunk 9]
```diff
@@ -2,4 +2,5 @@
 /// <reference path="../ts/index.d.ts" />
 export const $Cypress: Cypress.Cypress
 
-export default $Cypress
+export const $: typeof JQuery
```
Comment (elevatebart) [inline packages/driver/index.d.ts:5]: Why was this change necessary?
It seems not connected to the PR. It might be a dependency of the work.

Usually, this is supposed to be part of the jQuery types inside of your ``tsconfig,json`

File: packages/runner-shared/package.json [hunk 10]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@packages/runner-shared",
+  "version": "0.0.0-development",
+  "private": true,
+  "main": "src/index.ts",
+  "scripts": {
+    "test": "yarn test-unit",
+    "test-unit": "mocha --config test/.mocharc.json src/**/*.spec.* --exit"
```
Comment (elevatebart) [inline packages/runner-shared/package.json:8]: nit: Could we have used Cypress Component Testing for that ?

Comment (lmiller1990) [reply packages/runner-shared/package.json:8]: Maybe one day, but there is already 200+ tests written with enzyme, good to keep them for now.

File: packages/driver/index.d.ts (hunk 9 above)
Comment (lmiller1990) [reply packages/driver/index.d.ts:5]: Many places in the codebase do `import { $} from '@packages/driver'`. Until now they were all in JS files, so no-one noticed but the types were wrong. I added some TS files in this PR, revealing the wrong typings. So I fixed the types.

**Comment by lmiller1990** [comment]
@flotwig @elevatebart @jennifer-shehane addressed all issues - any outstanding blockers or further changes you'd like to see?

File: packages/runner-shared/src/index.ts (hunk 8 above)
Comment (flotwig) [reply packages/runner-shared/src/index.ts]: you have convinced me that it's not worth changing, at least :smile:

**Review by flotwig — APPROVED** [review]
